File: archived/src/python/ch01/1.1/l2/exercises_starter.py
import math
import sys
from ieee754_utils import (
    decompose, classify, float_to_bits, bits_to_float,
    exponent_unbiased, is_subnormal
)

def exercise_1_bits_of_one():
    """Return (sign, exponent_raw, mantissa) for 1.0 and unbiased exponent."""
    s, e, m = decompose(1.0)
    return s, e, m, exponent_unbiased(e)

def exercise_2_machine_epsilon_near_one():
    """Compute ULP(1.0) via nextafter and compare to sys.float_info.epsilon."""
    nxt = math.nextafter(1.0, math.inf)
    return nxt - 1.0, sys.float_info.epsilon

def exercise_3_classify_values():
    """Classify representative values."""
    values = [0.0, 1.0, float('inf'), float('nan'), 5e-324]
    return [(v, classify(v)) for v in values]

def exercise_4_roundtrip_bits(x: float):
    """Convert float to bits and back, return recovered value."""
    bits = float_to_bits(x)
    y = bits_to_float(bits)
    return bits, y

def exercise_5_find_first_subnormal():
    """Find the first subnormal above 0.0 using nextafter."""
    x = math.nextafter(0.0, 1.0)
    return x, is_subnormal(x)

def _demo():
    print("Ex1:", exercise_1_bits_of_one())  # expect sign=0, unbiased exponent=0, mantissa=0
    print("Ex2 ULP vs epsilon:", exercise_2_machine_epsilon_near_one())
    print("Ex3 classify:", exercise_3_classify_values())
    bits, y = exercise_4_roundtrip_bits(3.5)
    print(f"Ex4 roundtrip 3.5: bits=0x{bits:016x}, back={y}")
    print("Ex5 first subnormal > 0.0:", exercise_5_find_first_subnormal())

if __name__ == "__main__":
    _demo()

File: archived/src/go/ch01/1.1/l2/exercises_starter.go
package main

import (
	"fmt"
	"math"
)

func ex1BitsOfOne() (sign uint64, expRaw uint64, mant uint64, unbiased int64) {
	f := Decompose(1.0)
	return f.Sign, f.ExponentRaw, f.Mantissa, UnbiasedExponent(f.ExponentRaw)
}

func ex2ULPNearOne() (ulp float64, nextUp float64) {
	next := math.Nextafter(1.0, math.Inf(+1))
	return next - 1.0, next
}

func ex3Classify() [][2]string {
	vals := []float64{0.0, 1.0, math.Inf(+1), math.NaN(), math.SmallestNonzeroFloat64}
	out := make([][2]string, 0, len(vals))
	for _, v := range vals {
		out = append(out, [2]string{fmt.Sprintf("%.17g", v), kindToString(Classify(v))})
	}
	return out
}

func ex4Roundtrip(x float64) (bits uint64, y float64) {
	b := Float64ToBits(x)
	return b, BitsToFloat64(b)
}

func ex5BuildInfAndNaN() (inf float64, nanEqual bool) {
	finf := FPFields{Sign: 0, ExponentRaw: (1<<ExpBits - 1), Mantissa: 0}
	fnan := FPFields{Sign: 0, ExponentRaw: (1<<ExpBits - 1), Mantissa: 1}
	inf = Build(finf)
	nan := Build(fnan)
	return inf, (nan == nan)
}

func runExercises() {
	// show exact decimal demo too
	decimalExactDemo()

	s, e, m, ub := ex1BitsOfOne()
	fmt.Printf("Ex1: sign=%d exp_raw=%d mantissa=0x%013x (unbiased=%d)\n", s, e, m, ub)

	ul, next := ex2ULPNearOne()
	fmt.Printf("Ex2: ULP(1.0)=%.17g nextUp=%.17g\n", ul, next)

	for _, pair := range ex3Classify() {
		fmt.Printf("Ex3: %s -> %s\n", pair[0], pair[1])
	}

	bits, y := ex4Roundtrip(3.5)
	fmt.Printf("Ex4: 3.5 bits=0x%016x back=%.17g\n", bits, y)

	inf, nanEq := ex5BuildInfAndNaN()
	fmt.Printf("Ex5: built +inf=%g, built nan==nan? %v\n", inf, nanEq)
}

func kindToString(k FPKind) string {
	switch k {
	case FPNan:
		return "nan"
	case FPInf:
		return "inf"
	case FPZero:
		return "zero"
	case FPSubnormal:
		return "subnormal"
	default:
		return "normal"
	}
}
